src/CHIPBackend.cc

```diff
@@ -609,6 +615,40 @@ CHIPQueue *CHIPDevice::getActiveQueue() { return chip_queues[0]; }
 //*************************************************************************************
 CHIPContext::CHIPContext() {}
 CHIPContext::~CHIPContext() {}
+
+void CHIPContext::syncQueues(CHIPQueue *target_queue) {
+  logDebug("CHIPContext::syncQueues()");
+  std::vector<CHIPQueue *> queues = getQueues();
+  std::vector<CHIPQueue *> blocking_queues;
+
+  // Default queue gets created add init - always 0th in queue list
+  CHIPQueue *default_queue = queues[0];
+  queues.erase(queues.begin());
+
+  for (auto &q : queues)
+    if (q->getQueueType() == CHIPQueueType::Default)
+      blocking_queues.push_back(q);
+  logDebug("Num blocking queues: {}", blocking_queues.size());
+
+  // default stream waits on all non-blocking streams to complete
+  std::vector<CHIPEvent *> events_to_wait_on;
+  CHIPEvent *signal;
+  /**
+   * TODO:
+   * CHIPQueue->LastEvent must be initialized to marker event upon queue
+   * construction
+   */
+  if (target_queue == default_queue) {
+    for (auto &q : blocking_queues) events_to_wait_on.push_back(q->LastEvent);
+    signal = target_queue->enqueueBarrier(&events_to_wait_on);
+    target_queue->LastEvent = signal;  // TODO: replace with updateLastEvent()
+  } else {  // blocking stream must wait until default stream is done
```

Comment:
Missing test for nonBlocking queues, those only need to be synched with themselves

Reply:
follow up via slack